### PR TITLE
Add feature events

### PR DESCRIPTION
**Pull request by Jesus89**
Related to https://github.com/CartoDB/renderer-prototype/issues/223

**Add**

- `featureClick` tests
- `featureClickOut` event (+tests)
- `featureEnter` event (+tests)
- `featureLeave` event (+tests)
- `featureHover` event (+tests)

**Feature hover**

feature-hover

**Feature enter/leave**

feature-enter-leave

NOTE: featureIDs depend on the tiles, so if a feature covers different tiles (polygon) it will have different featureIDs in each tile.

### CODE REVIEW

File: src/core/dataframe.js [hunk 1]
```diff
@@ -117,6 +117,7 @@ export default class Dataframe {
             const inside = pointInTriangle(p, v1, v2, v3);
             if (inside) {
                 features.push({
+                    id: featureID,
```
Comment (davidmanzanares) [inline src/core/dataframe.js]: Related to https://github.com/CartoDB/renderer-prototype/issues/120

Comment (davidmanzanares) [reply src/core/dataframe.js]: I think we should merge this without waiting for this fix

File: examples/interactivity/feature-click.html [hunk 2]
```diff
@@ -87,6 +87,10 @@ <h2></h2>
       const interactivity = new carto.Interactivity([layer]);
       interactivity.on('featureClick', event => {
         updateCountry(event);
+        console.log('CLICK', event.features);
```
Comment (davidmanzanares) [inline examples/interactivity/feature-click.html]: Remove this?

File: examples/interactivity/feature-click.html [hunk 3]
```diff
@@ -87,6 +87,10 @@ <h2></h2>
...
+      });
+      interactivity.on('featureClickOut', event => {
+        console.log('CLICK OUT', event.features);
```
Comment (davidmanzanares) [inline examples/interactivity/feature-click.html]: same

File: src/api/interactivity.js [hunk 4]
```diff
@@ -65,6 +108,25 @@ export default class Interactivity {
...
     }
+    
+    /**
+     * Return the difference between the feature arrays A and B.
+     * The output value is also an array of features.
+     */
+    _getDiffFeatures(featuresA, featuresB) {
+        let featuresDiff = [];
+        const IDs = this._getFeatureIDs(featuresB);
+        for (let feature of featuresA) {
```
Comment (davidmanzanares) [inline src/api/interactivity.js]: I think this can be simplified by a `filter()`

File: examples/interactivity/feature-click.html (hunk 2 above)
Comment (Jesus89) [reply examples/interactivity/feature-click.html]: Yes. This is only for development purposes. Will be removed this week.

File: examples/interactivity/feature-click.html (hunk 3 above)
Comment (Jesus89) [reply examples/interactivity/feature-click.html]: same ^

File: src/api/interactivity.js (hunk 4 above)
Comment (Jesus89) [reply src/api/interactivity.js]: Yep, of course!

File: test/integration/user/api/interactivity.test.js [hunk 5]
```diff
@@ -71,6 +71,172 @@ describe('Interactivity', () => {
...
+                        expect(event.features[0]).toEqual({ id: 0, layerId: 'layer', properties: { cartodb_id: 1 } });
+                        done();
+                    });
+                    // Move mouse inside a feature
+                    map.fire('mousemove', { lngLat: { lng: 10, lat: 10 } });
+                });
+                layer.addTo(map);
+            });
+            
+            it('should fire a featureEnter event with a features list containing the entered feature', done => {
```
Comment (IagoLast) [inline test/integration/user/api/interactivity.test.js]: @Jesus89 I think this tests is duplicated :P

**Pull request merged**